### PR TITLE
Improving list data

### PR DESCRIPTION
**Pull request by robertbastian**
Previously, list data did internal deduplication between narrow/short/wide patterns in serde only. However, as wide patterns differ more between languages, that was blocking data-struct-level deduplication, so baked data basically didn't deduplicate. Splitting narrow/short/wide into different data structs only adds ~10% more data structs (instead of the expected 200%), and some additional lookup data.

### CODE REVIEW

File: provider/data/list/fingerprints.csv [hunk 1]
```diff
@@ -1,386 +1,1005 @@
...
-list/unit@2, vi, 1364B, -> as
-list/unit@2, xnr, 1366B, 5f6f58501e522c14
-list/unit@2, yrl, 1384B, fcc6d6177fc86427
-list/unit@2, yue, 1352B, -> ja
-list/unit@2, yue-Hans, 1352B, -> ja
-list/unit@2, zh, 1344B, 16c295fb252e0c3
-list/unit@2, zh-Hant, 1352B, -> ja
-list/unit@2, zu, 1367B, 6202988c3f3e90c9
+list/and@2, <lookup>, 1916B, 355 identifiers
+list/and@2, <total>, 39389B, 127 unique payloads
```
Comment (sffc) [inline provider/data/list/fingerprints.csv:2]: Praise: `176692B` to `39389B`

File: provider/data/list/fingerprints.csv [hunk 2]
```diff
@@ -1,386 +1,1005 @@
...
+list/and@2, af/N, 290B, d2846cc3cf177584
+list/and@2, af/S, 288B, abf92122ab812de
+list/and@2, af/W, 288B, -> af/S
```
Comment (sffc) [inline provider/data/list/fingerprints.csv:5]: Discussion: you use `N`, `S`, `W` here; I think Younies or someone was using `l` and `s`, and in datetime we use numbers `1`, `3`, `4` (2 means 2-digit and isn't part of the data marker attribute). Should we be more consistent? (Doesn't block this PR)

File: components/list/src/provider/mod.rs [hunk 3]
```diff
@@ -74,31 +73,36 @@ pub const MARKERS: &[DataMarkerInfo] = &[
...
+    /// and end with `{1}`, so we just store the string in between.
+    #[cfg_attr(feature = "serde", serde(borrow))]
+    pub middle: Cow<'data, str>,
+    /// The end pattern
+    #[cfg_attr(feature = "serde", serde(borrow))]
+    pub end: ConditionalListJoinerPattern<'data>,
+    /// The pair pattern, if it's different from the end pattern.
+    #[cfg_attr(feature = "serde", serde(borrow))]
+    pub pair: Option<ConditionalListJoinerPattern<'data>>,
+}
```
Comment (sffc) [inline components/list/src/provider/mod.rs:91]: Thought: we might consider changing this to use `Pattern`

**Review by sffc — APPROVED** [review]
Praise: Nice work and size improvement

**Review by Manishearth — APPROVED** [review]

File: provider/data/list/fingerprints.csv (hunk 2 above)
Comment (Manishearth) [reply provider/data/list/fingerprints.csv:5]: i'd prefer lowercase here, this feels like cardinal directions

extremely light preference, and very much non blocking

File: components/list/src/provider/mod.rs (hunk 3 above)
Comment (robertbastian) [reply components/list/src/provider/mod.rs:91]: I tried, pattern needs work to be usable, and it doesn't save much compared to this

**Pull request merged**